File: p135/python/p135.py

class Node:
    def __init__(self, value, left=None, right=None):
        self.value = value
        self.left = left
        self.right = right

    def __str__(self):
        return f'<Node: value={self.value}, left={self.left}, right={self.right}>'


def min_path_sum(root):
    m = float('inf')
    q = [(root, root.value)]

    while q:
        node, curr_sum = q.pop()

        if not node.left and not node.right:
            if m > curr_sum:
                m = curr_sum
            continue

        if node.left:
            q.append((node.left, curr_sum + node.left.value))
        if node.right:
            q.append((node.right, curr_sum + node.right.value))

    return m

File: p135/go/p135.go
package main

import "fmt"
import "math"

type Node struct {
	Value int
	Left  *Node
	Right *Node
}

type NodeWithSum struct {
	Node *Node
	Sum  int
}

func (n *Node) String() string {
	return fmt.Sprintf("<Node value=%d, left=%v, right=%v>", n.Value, n.Left, n.Right)
}

func minPathSum(root *Node) int {
	var queue []*NodeWithSum
	queue = append(queue, &NodeWithSum{
		Node: root,
		Sum:  root.Value,
	})
	min := math.MaxInt64

	for len(queue) > 0 {
		nodeWithSum := queue[0]
		queue = queue[1:]

		node := nodeWithSum.Node
		currSum := nodeWithSum.Sum

		if node.Left == nil && node.Right == nil {
			if currSum < min {
				min = currSum
			}
			continue
		}

		if node.Left != nil {
			queue = append(queue, &NodeWithSum{
				Node: node.Left,
				Sum:  currSum + node.Left.Value,
			})
		}
		if node.Right != nil {
			queue = append(queue, &NodeWithSum{
				Node: node.Right,
				Sum:  currSum + node.Right.Value,
			})
		}
	}

	return min
}
